File: internal/ingest/xml.go
package ingest

import (
	"html"
	"strconv"
	"strings"

	"gingest-desktop/internal/model"
)

func BuildFullXML(response model.GingestResponse) string {
	var sb strings.Builder

	sb.WriteString("<project_summary>\n")
	sb.WriteString("Project: ")
	sb.WriteString(response.ProjectName)
	sb.WriteString("\n")
	sb.WriteString("Total Files: ")
	sb.WriteString(strconv.Itoa(response.FileCount))
	sb.WriteString("\n")
	sb.WriteString("Estimated Tokens: ")
	sb.WriteString(strconv.FormatInt(response.EstimatedTokens, 10))
	sb.WriteString("\n")
	sb.WriteString("</project_summary>\n\n")

	sb.WriteString("<directory_tree>\n")
	sb.WriteString(".\n")
	sb.WriteString(GenerateTreeText(response.DirectoryTree, ""))
	sb.WriteString("</directory_tree>\n\n")

	sb.WriteString("<files>\n")
	appendFiles(&sb, response.DirectoryTree)
	sb.WriteString("</files>")

	return sb.String()
}

func GenerateTreeText(nodes []model.TreeNode, prefix string) string {
	var sb strings.Builder

	for i, node := range nodes {
		isLast := i == len(nodes)-1
		connector := "├── "
		childPrefix := prefix + "│   "

		if isLast {
			connector = "└── "
			childPrefix = prefix + "    "
		}

		sb.WriteString(prefix)
		sb.WriteString(connector)
		sb.WriteString(node.Label)

		if !node.IsFile {
			sb.WriteString("/")
		}

		sb.WriteString("\n")

		if len(node.Children) > 0 {
			sb.WriteString(GenerateTreeText(node.Children, childPrefix))
		}
	}

	return sb.String()
}

func appendFiles(sb *strings.Builder, nodes []model.TreeNode) {
	for _, node := range nodes {
		if node.IsFile {
			sb.WriteString(`<file path="`)
			sb.WriteString(html.EscapeString(node.FullPath))
			sb.WriteString(`">`)
			sb.WriteString("\n")
			sb.WriteString("<![CDATA[\n")
			sb.WriteString(safeCDATA(node.Content))
			sb.WriteString("\n]]>\n")
			sb.WriteString("</file>\n\n")
			continue
		}

		if len(node.Children) > 0 {
			appendFiles(sb, node.Children)
		}
	}
}

func safeCDATA(content string) string {
	return strings.ReplaceAll(content, "]]>", "]]]]><![CDATA[>")
}

File: frontend/src/utils/xml.ts
import type { TreeNode } from '../types/gingest'

export const escapeXmlAttribute = (value: string) => {
    return value
        .replaceAll('&', '&amp;')
        .replaceAll('"', '&quot;')
        .replaceAll('<', '&lt;')
        .replaceAll('>', '&gt;')
}

export const safeCDATA = (value: string) => {
    return value.replaceAll(']]>', ']]]]><![CDATA[>')
}

export const generateTreeText = (nodes: TreeNode[], prefix = ''): string => {
    let text = ''

    nodes.forEach((node, index) => {
        const isLast = index === nodes.length - 1
        const connector = isLast ? '└── ' : '├── '
        const childPrefix = prefix + (isLast ? '    ' : '│   ')

        text += prefix + connector + node.label + (node.isFile ? '' : '/') + '\n'

        if (node.children && node.children.length > 0) {
            text += generateTreeText(node.children, childPrefix)
        }
    })

    return text
}

export const buildXmlByFiles = (
    projectName: string,
    directoryTree: TreeNode[],
    files: TreeNode[],
    exportType: string,
) => {
    const estimatedTokens = Math.floor(
        files.reduce((sum, file) => sum + (file.content?.length || 0), 0) / 4,
    )

    let xml = ''
    xml += '<project_summary>\n'
    xml += `Project: ${projectName}\n`
    xml += `Export Type: ${exportType}\n`
    xml += `Total Files: ${files.length}\n`
    xml += `Estimated Tokens: ${estimatedTokens}\n`
    xml += '</project_summary>\n\n'

    xml += '<directory_tree>\n'
    xml += '.\n'
    xml += generateTreeText(directoryTree)
    xml += '</directory_tree>\n\n'

    xml += '<files>\n'

    files.forEach((file) => {
        xml += `<file path="${escapeXmlAttribute(file.fullPath || file.label)}">\n`
        xml += '<![CDATA[\n'
        xml += safeCDATA(file.content || '')
        xml += '\n]]>\n'
        xml += '</file>\n\n'
    })

    xml += '</files>'

    return xml
}

export const buildSuggestedXmlFileName = (projectName: string, selected: boolean) => {
    if (!projectName) return 'gingest_export.xml'

    const suffix = selected ? '_selected' : '_full'

    return (
        projectName
            .replace(/^Local:\s*/, '')
            .replace(/[\\/:*?"<>|]/g, '_') +
        `_gingest${suffix}.xml`
    )
}